tests/csapi/apidoc_sendtodevice_test.go

```diff
@@ -0,0 +1,196 @@
+package csapi_tests
+
+import (
+	"fmt"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/match"
+)
+
+func TestSendToDevice(t *testing.T) {
+	deployment := Deploy(t, b.BlueprintOneToOneRoom)
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	bob := deployment.Client(t, "hs1", "@bob:hs1")
+	charlie := deployment.RegisterUser(t, "hs1", "charlie", "$uperSecretPass", false)
+
+	// checks used to verify the sendToDevice messages are correct
+	checks := []match.JSON{
+		match.JSONKeyEqual("sender", alice.UserID),
+		match.JSONKeyEqual("type", "m.room_key_request"),
+		match.JSONKeyEqual("content.action", "request"),
+		match.JSONKeyEqual("content.requesting_device_id", alice.DeviceID),
+		match.JSONKeyEqual("content.request_id", "1"),
+	}
+
+	// sytest: Can recv a device message using /sync
+	t.Run("Can recv a device message using /sync", func(t *testing.T) {
+		// Create a m.room_key_request sendToDevice message
+		reqBody := client.WithJSONBody(t, map[string]interface{}{
+			"messages": map[string]interface{}{
+				bob.UserID: map[string]interface{}{
+					bob.DeviceID: map[string]interface{}{
+						"action":               "request",
+						"requesting_device_id": alice.DeviceID,
+						"request_id":           "1",
+					},
+				},
+			},
+		})
+		txnID := "1"
+		// sytest: Can send a message directly to a device using PUT /sendToDevice
+		alice.MustDoFunc(t, "PUT", []string{"_matrix", "client", "v3", "sendToDevice", "m.room_key_request", txnID}, reqBody)
+
+		// verify the results from /sync
+		nextBatch := bob.MustSyncUntil(t, client.SyncReq{}, verifyToDeviceResp(t, bob, checks, 1))
+		// clear previously received sendToDevice events
+		bob.MustSyncUntil(t, client.SyncReq{Since: nextBatch}, verifyToDeviceResp(t, bob, checks, 0))
```

Comment:
Using `MustSyncUntil` would be the right thing to do here, as there is no guarantee that the to-device message will be delivered on the first /sync, even if the PUT 200 OKs (it could be OKd after persistence but before downstream workers/components for the sync API are aware of it).

Reply:
>  there is no guarantee that the to-device message will be delivered on the first /sync

but `verifyToDeviceResp(t, bob, checks, 0)` checks that no to-device messages are delivered in the `/sync`. So this is allowing an unexpected/repeated to-device event before getting to that state. I don't see why we would explicitly want to allow to-device events that we don't want.